Iterable tapes: allow `for op in tape`

antalszava opened the pull request:
**Context**

`QuantumTape` objects record quantum operations and measurements. These can be accessible by querying the `operations` and `measurements` attributes.

At times we'd like to easily access the underlying circuit and/or all of the operators.

**Changes**

Implements the requirements for making a `QuantumTape`:
1. An iterable (https://docs.python.org/3/glossary.html#term-iterable);
2. And a sequence (https://docs.python.org/3/glossary.html#term-sequence).

Furthermore, changes relevant transforms to use the `for op in tape` syntax.

**Benefits**

Better usability of tapes.

**Possible drawbacks**
N/A

**Related issues**
N/A

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -515,6 +527,8 @@ def _update(self):
         self._update_trainable_params()
         self._update_observables()
 
+        self._circuit = self.operations + self.measurements
```
I think you should define an empty circuit in __init__ it should solve some failures in the test suite.

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -349,6 +349,18 @@ def __exit__(self, exception_type, exception_value, traceback):
+        self._iter_idx = 0
+        return self
+
+    def __next__(self):
```
Nice and clean solution in my opinion! It is great to make this object as an iterable, it is more user friendly. This type of addition should make it `Tape` an easier object to use!

albi3ro commented on tests/tape/test_tape.py:
```diff
@@ -273,6 +273,111 @@ def test_repr(self):
+
+        assert len(tape) == len(expected)
+
+    def test_tape_as_list(self, make_tape):
```
Doesn't look like it actually gets converted to a list here. Should this be testing `list(tape)`?

antalszava replied:
Good catch! Updated.

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -236,8 +236,11 @@ class QuantumTape(AnnotatedQueue):
+    Once constructed, the tape may act as a quantum circuit and information
+    about the quantum circuit can be queried:
 
+    >>> list(tape)
```
Great that we can convert it so easily!

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -249,6 +252,19 @@ class QuantumTape(AnnotatedQueue):
+    RX(0.133, wires=['a'])
+    expval(PauliZ(wires=[0]))
+
+    Tapes also support indexing.
```
I might be worth to add an example here.

antalszava replied:
Good idea, added!

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -349,6 +365,42 @@ def __exit__(self, exception_type, exception_value, traceback):
+
+        The circuit is created with the assumptions that:
+
+        * The ``operations`` attribute contains quantum operations and
```
It is good to give this information. Do you think it would be desirable to defer all measurement to the end for the tape?

antalszava replied:
>  Do you think it would be desirable to defer all measurement to the end for the tape?

That's what happens with the `defer_measurements` transform, right?

In terms of whether or not `tape.measurements` should contain both mid-circuit measurements and terminal measurements: that's a bit more complex (and likely out of scope for here). For now, it seems that this "hack" works well and we would not gain much from doing so in development.

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -349,6 +365,42 @@ def __exit__(self, exception_type, exception_value, traceback):
+        """
+        return self.operations + self.measurements
+
+    def __iter__(self):
+        """list[.Operator, .MeasurementProcess]: Return an iterator to the
+        underlying quantum circuit object."""
+        return iter(self.circuit)
+
+    def __getitem__(self, idx):
+        """list[.Operator]: Return the indexed operator from underlying quantum
+        circuit object."""
+        return self.circuit[idx]
+
+    def __len__(self):
+        """int: Return the number of operators in the underlying quantum
+        circuit object."""
+        return len(self.circuit)
```
Clean and clear implementation!

rmoyard commented on pennylane/tape/tape.py:
```diff
@@ -349,6 +365,42 @@ def __exit__(self, exception_type, exception_value, traceback):
+        return self.circuit[idx]
+
+    def __len__(self):
+        """int: Return the number of operators in the underlying quantum
```
Might be better to say the sum of number of operations and observables?

albi3ro replied:
+1

rmoyard commented on tests/tape/test_tape.py:
```diff
@@ -273,6 +273,139 @@ def test_repr(self):
+
+        assert len(tape) == len(expected)
+
+    def test_iteration_preserves_circuit(self):
```
Valuable test, I was also wondering of it could affect it!

antalszava replied:
Oh cool! I was wondering if it's worth adding... :thinking: Good to have now though. :+1:

rmoyard reviewed (commented):
Great job @antalszava, I've left some comments!

albi3ro commented on tests/tape/test_tape.py:
```diff
@@ -273,6 +273,139 @@ def test_repr(self):
         expected = "<QuantumTape: wires=[0], params=1>"
         assert s == expected
 
+    def test_circuit(self):
```
```suggestion
    def test_circuit_property(self):
```

antalszava commented:
[sc-16250]

antalszava commented:
Thanks for the review and comments @rmoyard @albi3ro! I've addressed the points raised.

If there are any other cases that could be worth testing and come to mind, let me know.

albi3ro commented on doc/releases/changelog-dev.md:
```diff
@@ -18,6 +18,37 @@
 
 <h3>Improvements</h3>
 
+* `QuantumTape` objects are now iterable and can be used in `for` loops or
```
I think important to also mention that we can index into them via `tape[i]`.

antalszava replied:
Sure thing! Updated.

albi3ro reviewed (approved):
Left one optional comment about the changelog, but thrilled this is finally happening! 💯
